[8.0][event_product] New module to combine events and products

yajo opened the pull request:
Following discussion from #14, and aiming for https://github.com/OCA/l10n-spain/issues/174, this module links events and products.

This implementation follows [what was discussed before](https://docs.google.com/document/d/1WiOB1yAWTR2aROpM6O632QWXwgG4hLRUpTHU2Pv4yEQ/edit?usp=sharing), and of course this is the first of a collection of modules to implement training in Odoo.

See README for instructions, please.

CC @pedrobaeza @elicoidal @dreispt @rafaelbn.

Thanks!

Commit:
Add module event_product.

dreispt commented on event_product/README.rst:
```diff
@@ -0,0 +1,119 @@
+any product as long as:
+
+* The product's *Is an event* checkbox is enabled.
+* The product's event type is empty, or matches the event's.
```
s/product's/Product

yajo replied:
Are you sure no `'s` is needed? :confused:

dreispt replied:
Reading again - you're right, it's OK.

dreispt commented on event_product/exceptions.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class TypeMismatchError(EventProductValidationError):
+    __doc__ = _("Product has type {product_type}, but event has {event_type}.")
```
My personal opinion is that these custom exceptions do not go with the _de facto_ Odoo development standards, and increases the codebase for no visible benefit. 
But again, it's just my opinion.

yajo replied:
This does not come only from Odoo experience, but from outer too. Tests are also "more code to maintain" but serve the same purpose: make code less error-prone.

It's fun that you mention it, because this time was one of those where the test suite revealed a bug that would have passed silently without explicit exceptions.

dreispt replied:
You mean, you about a wrong error message being raised?

elicoidal replied:
s/but/whereas

yajo replied:
@dreispt Yes, because some parts of code were being hit before anothers. I know I could do that with error message and testing that message in the test, but if after all I need to add more code, I prefer it this way.

dreispt replied:
@Yajo You could achieve just the same if using only plain Odoo exceptions. I still can't see any benefit on that - IMO it's more code for no advantage. But I block anything because of that.

dreispt commented:
:+1: (code review, no test)

dreispt commented on event_product/README.rst:
```diff
@@ -0,0 +1,119 @@
+  in products, which is added by the *event_sale* module, which installs
+  *sale*; so unless Odoo divides that module in smaller parts with smaller
+  dependencies, the only workaround is to give permissions to use *sale* to
+  nobody.
```
Thanks!

elicoidal commented on event_product/README.rst:
```diff
@@ -0,0 +1,119 @@
+==============
+
+This module extends the functionality of events and products to support
+combining them and allow you to have a *catalog* of events that you organize.
```
Current sentence is weird and not so clear at first. What about:
This module extends the functionality of events and products allowing their combination and the organization of a _catalog_ of events.

elicoidal commented on event_product/README.rst:
```diff
@@ -0,0 +1,119 @@
+Known issues / Roadmap
+======================
+
+* If you install this module, you will install *sale* too. If you do not want
```
Installing this module will install _sale_ too

elicoidal commented on event_product/README.rst:
```diff
@@ -0,0 +1,119 @@
+======================
+
+* If you install this module, you will install *sale* too. If you do not want
+  to sale events, but just to handle internal ones, this can be disturbing
```
If you do not intend to sell events, but just handle internal ones, this can be annoying since the sale module requires complex configuration with accounting and external dependencies.

dreispt replied:
@elicoidal Yes, my point also. 
But we debated that point, and there are implementation details complicating that.
I believe that we can overcome them. So I agreed with the current design and asked for this note to be on the issues/roadmap.

yajo replied:
I think @elicoidal's point is about changing the phrasing. I'll fix your comments ASAP.

elicoidal replied:
@Yajo Correct. Actually I made a comment that disappeared about that ...
I am not sure: is there a specific convention to ask for rephrasing?
I use s/this/that for short changes but it is not handy for long sentences.
I could use:
"This is the new sentence"
or state Rephrase: "blabla"

yajo replied:
> "This is the new sentence"

Seems good.

elicoidal commented on event_product/README.rst:
```diff
@@ -0,0 +1,119 @@
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
https

elicoidal commented on event_product/__openerp__.py:
```diff
@@ -0,0 +1,26 @@
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "external_dependencies": {
```
Please remove empty keys

elicoidal commented on event_product/exceptions.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class EventAndTicketError(EventProductValidationError):
+    __doc__ = _("It cannot be a whole event and a ticket at the same time.")
```
Sorry I dont understand what is a "whole event" please rephrase

yajo replied:
It means this product refers to an event as a full product, not to a ticket for an event. You're more skilled in English, how could I say that?

elicoidal commented on event_product/exceptions.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class ProductEventTypeError(EventProductValidationError):
+    __doc__ = _("Product is already used in some event with a different type.")
```
s/some/another

elicoidal commented on event_product/models/product.py:
```diff
@@ -0,0 +1,81 @@
+
+    is_event = fields.Boolean(
+        "Is an event",
+        help="This product defines a whole event (NOT an event ticket).")
```
redefine "whole event"

Commit:
Remove unused exception.

Commit:
Spell checking, updated translation.

yajo commented:
All fixed. I've got the next module on hold until this one is merged (I do not want to confuse reviewers anymore).

dreispt commented:
@Yajo Please open a PR for it, just don't repeat this module in that PR. You can title it with `[WIP]`
The build will fail from the missing dependency, but no problem.

rafaelbn commented:
Tested, some questions:
- In readme you say
  
  > This module extends the functionality of events and products allowing their combination and the organization of a catalog of events.
  - If you have a catalog of events you would like to sell this catalog, isn't it?
  - How to create a SO from Event?
- If you make a SO with a product which is an Event, why don't create a event when confirming it?
- Go to reporting - Event - Event Analysis, must you add there the option of grouping by product?
- Must you add a smart-button in product.product for events? In Odoo when there is a relation of two models usually this created actions and short-cuts for quicker and better info

![2015-12-01_23-21-05](https://cloud.githubusercontent.com/assets/8736623/11516215/440f12ac-9882-11e5-9f1f-4f642173a10a.png)

I'm trying to see usefulness as you see.

dreispt commented:
@rafaelbn SO are out of scope for this module: that logic is to be implemented in a separate module.

rafaelbn commented:
OK @dreispt . Sure.

What about this two:
- Go to reporting - Event - Event Analysis, must you add there the option of grouping by product?
- Must you add a smart-button in product.product for events? In Odoo when there is a relation of two models usually this created actions and short-cuts for quicker and better info

Thanks

dreispt commented:
@Yajo, can you answer @rafaelbn ?

yajo commented:
@rafaelbn 

> If you have a catalog of events you would like to sell this catalog, isn't it?
> How to create a SO from Event?
> If you make a SO with a product which is an Event, why don't create a event when confirming it?

Those seem reasonable things to do by this module, but its main purpose for now is to serve as a base for training courses, as discussed in #14.

Unlike @dreispt, I think this module is a good place to put that. He wants to avoid `sale` dependency, but since it is not possible (see README's known issues), why not just use this module for that?

In any case, the best I can do for now is to add @rafaelbn's suggestions to roadmap.

> Go to reporting - Event - Event Analysis, must you add there the option of grouping by product?
> Must you add a smart-button in product.product for events? In Odoo when there is a relation of two models usually this created actions and short-cuts for quicker and better info

Good points. I'll fix those ASAP.

Commit:
Correct default values for new event products.

Commit:
Increase coverage and fix new bugs.

- Exceptions have new messages that can be understood when raised from
  product views.
- Exceptions support UTF-8 message formatting.
- Almost all product.product tests now test product.template too.

Commit:
Add event count button.

Commit:
Update known issues/roadmap.

yajo commented:
All fixed, including new bugs I found and known issues section with @rafaelbn comments.

Commit:
Hide event count for non-event products.

Commit:
Defaults when creating events from products link.

Commit:
Search views.

Commit:
Hide field for non-event users.

yajo commented:
Last commits improve UX hiding fields for non-event users and adding search views and context default for event users.

Commit:
Update Spanish translation.

elicoidal commented on event_product/README.rst:
```diff
@@ -0,0 +1,121 @@
+
+* Wedding.
+    * Variants:
+        * Until 50 attendees.
```
up to

elicoidal commented on event_product/README.rst:
```diff
@@ -0,0 +1,121 @@
+
+To create your *event products*, you need to:
+
+* Go to *Sales > Products > Products > Create*.
```
better would be enumerated lists

Commit:
Requested fixes in README.

yajo commented:
Thanks @elicoidal, fixed.

Commit:
Rename base model and class.

Commit:
Create a new "Events" tab.

Since this module serves as a base for many product-event modules, I
add this tab to give them a place to put their additional controls.

To make it more consistent with all those, I move the type of event
there.

Move event product filter to its place.

Merge product views in one.

Commit:
Easier to read constrain.

rafaelbn commented:
Great! :+1:
